### PR TITLE
Fixed camera_to_rays implementation, slightly modified interface

### PR DESCRIPTION
**Pull request by jackd**
Resolved [this issue](https://github.com/mikedh/trimesh/issues/486). In addition, made the following interface changes to `camera_to_rays` (and `Camera.to_rays`):

* `angles` is no longer an intermediate value, so is not calculated/returned.
* `origin` is no longer tiled to match `vectors`' shape.
* `vectors` in not longer `unitize`d. This means rays point to the image plane rather than lying on a sphere. This makes it easier to make frustums with `z_near` and `z_far` values. They can always be unitized afterwards by the caller.

These interface changes can be accepted/rejected independently - just thought I'd throw them all in the same PR for convenience, but happy to change them back.

~~I've had to introduce a slight [hack](https://github.com/jackd/trimesh/blob/release/major/examples/raytrace.py#L36) to get the raytrace example working. I have no idea why this is necessary... if that makes sense/can be justified by changes elsewhere feel free to fix things up, otherwise I'll look into it when I get a chance.~~

There's a slight hack [here](https://github.com/jackd/trimesh/blob/release/major/examples/voxel_silhouette.py#L24) to make things work. Feels a bit dirty - I'm just reordering the rays so the output silhouette matches the image - but I'm not sure why it's necessary...

### CODE REVIEW

File: trimesh/scene/cameras.py [hunk 1]
```diff
@@ -307,6 +318,23 @@ def look_at(points, fov, rotation=None, distance=None, center=None):
     return cam_pose
 
 
+def _ray_pixel_coords(camera):
```
Comment (mikedh) [inline trimesh/scene/cameras.py]: What's the reason to break `_ray_pixel_coords` out from `camera_to_rays`? Is it used elsewhere?

**Comment by mikedh** [comment]
Nice! I knew you'd get it in the end 😄. I had a quick API question which I left inline. 

Is there any simple check we could run in unit tests for camera-ray correspondence? Maybe check if everything is inside angle bounds, or something? Or if you were feeling fancy, create an optical plane of points and check that each vector comes within epsilon of its expected pixel location?

File: trimesh/scene/cameras.py (hunk 1 above)
Comment (jackd) [reply trimesh/scene/cameras.py]: in `Camera.angles`

Comment (mikedh) [reply trimesh/scene/cameras.py]: Ah gotcha. Any reason to not make it "public-ish"?

Comment (jackd) [reply trimesh/scene/cameras.py]: Happy to do so, I just don't think it's currently very well designed/named. I'm unsure if it'd be used much elsewhere, and whether people will know to find it even if it is. The description is something along the lines of

> the `xy` coordinates of pixels in the `z = -1` plane and the camera is looking along the `-z` axis. Reshaping to `res_x, res_y, 2` will give you the image coordinates except it's flipped up-down for some reason I don't understand.`

See comment below about flipping issue.

**Comment by jackd** [comment]
Not sure I quite follow what you mean in terms of tests - and before this gets accepted I'd really like to get to the bottom of the flipping issue. I'm not really a graphics person - and the reason I've tried to move a lot of my stuff towards this package is that I don't have to be :) . But I can't for the life of me work out why I need to keep flipping things up/down to make them consistent. Is it something to do with a front-back reflection, because the image plane is behind the camera or something? Differences in left-handed vs right-handed coordinate frames?

The folowing demonstrates the need to flip (and potentially a starting point for the tests you were mentioning above). I create a scene with default camera and put a pointcloud down at the ray endpoints (on the z=-1 plane) with dynamic coloring. We then compare renderings to the point cloud colors. Qualitatively the results match (so long as you include the mysterious flip). I could easily embed that flip into `_ray_pixel_coords`, but I'd like to understand why that's necessary so as not to introduce further bugs down the line...

```python

import numpy as np
import trimesh
from PIL import Image
import os

# bugged impl - cannot set_camera on empty scene
# scene = trimesh.Scene()
# scene.set_camera()

# create a sphere that allows us to get a camera that doesn't raise
sphere = trimesh.primitives.Sphere()
scene = sphere.scene()

origin, rays = scene.camera.to_rays(scene.camera_transform)
num_rays = len(rays)
endpoints = rays + origin

# create distinguishable colors
colors = trimesh.util.grid_linspace([[0, 0], [1, 1]], scene.camera.resolution)
colors = np.concatenate((colors, np.zeros_like(colors[..., :1])), axis=-1)
colors = (colors * 255).astype(np.uint8)

scene.add_geometry(trimesh.PointCloud(endpoints, colors=colors))
image = np.array(Image.open(trimesh.util.wrap_as_stream(
    scene.save_image(resolution=None))))
image = image[:, :, :3]

colors = np.reshape(colors, tuple(scene.camera.resolution) + (3,))
colors = colors[:, -1::-1]  # why is this necessary?
# make y-x ordering for comparison to image
colors = np.transpose(colors, (1, 0, 2))


def vis():
    import matplotlib.pyplot as plt
    _, ax = plt.subplots(1, 3)
    ax[0].imshow(colors)
    ax[0].set_title('colors flipped up-down')
    ax[1].imshow(image)
    ax[1].set_title('rendering')
    diff = np.abs(image.astype(int) - colors.astype(int))
    print('max difference: %s' % np.max(diff, axis=(0, 1)))
    diff = diff / np.max(diff)
    ax[2].imshow(diff)
    ax[2].set_title('rescaled difference')
    plt.show()


vis()
```

**Comment by jackd** [comment]
Sorry for the spam. After drawing some pictures and waving my right hand in a pistol-grip around for a bit, I'm now satisfied why I had to flip things around and have effectively moved that flip inside the now public `ray_pixel_coords` function. Also implemented changes I discussed [here](https://github.com/mikedh/trimesh/pull/469) (happy to roll-back if you don't like them).

Let me know if there's anything else needed for this.

**Comment by mikedh** [comment]
Looks great, thanks for all the work on voxels!! Just pressed merge :sweat_smile:

**Pull request merged**